explorer/src/providers/accounts/rewards.tsx

```diff
@@ -0,0 +1,196 @@
+import { reportError } from "utils/sentry";
+
+const PAGE_SIZE = 15;
+const timeout = (ms: number | undefined) =>
+  new Promise((resolve) => setTimeout(resolve, ms));
+
+export type Rewards = {
+  highestFetchedEpoch?: number;
+  lowestFetchedEpoch?: number;
+  rewards: (InflationReward | null)[];
+  foundOldest?: boolean;
+};
+
+export type RewardsUpdate = {
+  rewards: (InflationReward | null)[];
+  foundOldest?: boolean;
+};
+
+type State = Cache.State<Rewards>;
+type Dispatch = Cache.Dispatch<RewardsUpdate>;
+
+function reconcile(
+  rewards: Rewards | undefined,
+  update: RewardsUpdate | undefined
+): Rewards | undefined {
+  if (update === undefined) {
+    return rewards;
+  }
+
+  const combined = (rewards?.rewards || [])
+    .concat(update.rewards)
+    .filter((value) => value !== null);
+
+  const foundOldest = update.foundOldest;
+
+  return {
+    rewards: combined,
+    highestFetchedEpoch: combined[0]?.epoch,
+    lowestFetchedEpoch: combined[combined.length - 1]?.epoch,
+    foundOldest,
+  };
+}
+
+export const StateContext = React.createContext<State | undefined>(undefined);
+export const DispatchContext = React.createContext<Dispatch | undefined>(
+  undefined
+);
+
+type RewardsProviderProps = { children: React.ReactNode };
+
+export function RewardsProvider({ children }: RewardsProviderProps) {
+  const { url } = useCluster();
+  const [state, dispatch] = Cache.useCustomReducer(url, reconcile);
+
+  React.useEffect(() => {
+    dispatch({ type: ActionType.Clear, url });
+  }, [dispatch, url]);
+
+  return (
+    <StateContext.Provider value={state}>
+      <DispatchContext.Provider value={dispatch}>
+        {children}
+      </DispatchContext.Provider>
+    </StateContext.Provider>
+  );
+}
+
+async function fetchRewards(
+  dispatch: Dispatch,
+  pubkey: PublicKey,
+  cluster: Cluster,
+  url: string,
+  fromEpoch?: number,
+  highestEpoch?: number
+) {
+  dispatch({
+    type: ActionType.Update,
+    status: FetchStatus.Fetching,
+    key: pubkey.toBase58(),
+    url,
+  });
+
+  const connection = new Connection(url);
+
+  if (!fromEpoch && highestEpoch) {
+    fromEpoch = highestEpoch;
+  }
+
+  if (!fromEpoch) {
+    try {
+      const epochInfo = await connection.getEpochInfo();
+      fromEpoch = epochInfo.epoch - 1;
+    } catch (error) {
+      if (cluster !== Cluster.Custom) {
+        reportError(error, { url });
+      }
+
+      return dispatch({
+        type: ActionType.Update,
+        status: FetchStatus.FetchFailed,
+        key: pubkey.toBase58(),
+        url,
+      });
+    }
+  }
+
+  const getInflationReward = async (epoch: number) => {
+    try {
+      const result = await connection.getInflationReward([pubkey], epoch);
+      return result[0];
+    } catch (error) {
+      if (cluster !== Cluster.Custom) {
+        reportError(error, { url });
+      }
+    }
+    return null;
+  };
+
+  let foundSome = false;
+  let results;
+  while (!foundSome && fromEpoch >= 0) {
```

Comment:
Doesn't feel great that this could be fetching rewards for hundreds of epochs. I think these requests would consume quite a bit of bandwidth for bigtable. How about just fetching one page at a time?